186719432 open docs from sort work tab

dennisrcao opened the pull request:
Fulfills this ticket
[**[MultiDoc]** Open docs from Sort Work tab
](https://www.pivotaltracker.com/n/projects/2441242/stories/186719432)

Also all documents in sort work view should have the name prepended in the thumbnail document title.

Deployed link:
https://collaborative-learning.concord.org/branch/open-docs-from-sort-work-tab/?appMode=demo&demoName=dennis1&fakeClass=1&fakeUser=teacher:1&problem=1.1&unit=example-config-subtabs&curriculumBranch=sort-tab-dev-3

scytacki commented on src/components/document/sort-work-view.tsx:
```diff
@@ -1,67 +1,179 @@
+
+  const tabState = navTabSpec && persistentUI.tabs.get(navTabSpec?.tab);
+  const openDocumentsLength = tabState?.openDocuments.size;
+  const showSortWorkDocumentArea = openDocumentsLength && (openDocumentsLength > 1);
```
Shouldn't this be:
```
  const showSortWorkDocumentArea = openDocumentsLength && (openDocumentsLength > 0);
```
Or perhaps even more simple:
```
  const showSortWorkDocumentArea = !!openDocumentsLength;
```

scytacki commented on src/components/document/sort-work-view.tsx:
```diff
@@ -1,67 +1,179 @@
 import "./sort-work-view.scss";
 
-export const SortWorkView:React.FC = observer(function SortWorkView(){
+interface IProps {
+  tabSpec: NavTabModelType
+}
```
I'd guess you don't really need this. The only thing I think you are using it for is the string "sort-work", which will always be the same for this component.

scytacki commented on src/components/navigation/nav-tab-panel.tsx:
```diff
@@ -55,9 +56,20 @@ export class NavTabPanel extends BaseComponent<IProps> {
                 { tabs?.map((tabSpec, index) => {
                     const tabClass = `top-tab tab-${tabSpec.tab}
                                       ${selectedTabIndex === index ? "selected" : ""}`;
+                    const handleSortWorkClick = () => {
+                      closeSubTabDocument("sort-work", "sort-work");
```
this is probably another way to do this which is closer to how the other panels work, but I haven't looked.

dennisrcao commented on src/models/stores/nav-tabs.ts:
```diff
@@ -10,7 +10,7 @@ export const NavTabsConfigModel = types
   })
   .views(self => ({
-    getNavTabSpec(tabId: ENavTab) {
+    getNavTabSpec(tabId: ENavTab | string) {
```
I had to modify this so that "sort-work" could be accepted.

scytacki replied:
Rather than doing this, you should update ENavTab to include "sort-work".

dennisrcao replied:
Turns out `ENavTab` is an exported `enum` so I used `ENavTab.kSortWork`

dennisrcao commented on src/components/document/sort-work-view.scss:
```diff
@@ -1,21 +1,145 @@
+    }
+  }
+
+  .focus-document {
```
So the CSS was basically copied over from `section-document-or-browser.scss` 
So I was thinking since the "My Work" and "Class Work" tab only use `<SectionDocumentOrBrowser>`, and since we are phasing those out (correct me if I'm wrong - I believe you mentioned Leslie saying that). I just figured I'd copy it over  - I mean I figure we'd eventually get rid of `section-document-or-browser.scss`

dennisrcao replied:
We'd  also eventually want to move this CSS into a common file that all Nav Tabs that open up thumbnail documents could point to since these are children of parent classes with prefix "sort-work" etc.

scytacki replied:
The best refactoring in the future would be to extract a component from DocumentView which just renders the document with the title. And then DocumentView can overlay its paging stuff on top of that. But that can be for another time when this has settled down.  

I'm not sure if we'll be getting rid of "My Work", but I have heard that "Class Work" will go away if the new Sort Work tab works well for users.

dennisrcao commented:
@scytacki  I left some comments above - there's probably still stuff I could clean up but let me know your thoughts so far.

scytacki reviewed (changes requested):
This is looking good. I left a couple of inline comments.

And here are couple of general comments:
- it'd be good to add a cypress test which just opens this tab and switches between the two types of sort, and just opens a document. It doesn't need to test anything extensive, just make sure CLUE doesn't crash when opening this tab. QA can fill this test out further to make sure things are working properly. This simple test should increase the code coverage of this PR above the threshold.
- Update the branch and the base of the PR. This way the PR is only showing the stuff that you've changed instead of including the sort-by-name-and-group changes. Details below.

### Updating the branch base of the PR
1. Merge the branch `186506848-sort-by-name-and-group` from the other PR:
https://github.com/concord-consortium/collaborative-learning/pull/2113
into this PRs branch `186719432-open-docs-from-sort-work-tab`.
2. Set the base of this PR to `186506848-sort-by-name-and-group` that way it will only only show the new changes. 

Perhaps by the time you start for the day, https://github.com/concord-consortium/collaborative-learning/pull/2113 will be merged to master so in that case you just need to merge master into this PR.

scytacki commented on src/components/document/sort-work-document-area.tsx:
```diff
@@ -0,0 +1,91 @@
+import { getDocumentDisplayTitle } from "../../models/document/document-utils";
+import EditIcon from "../../clue/assets/icons/edit-right-icon.svg";
+import { observer } from "mobx-react";
+import { tabName } from "./sort-work-view";
```
This is a circular reference. sort-work-view is importing sort-work-document-area to be able to render the SortWorkDocumentArea component. And then here sort-work-document-area is importing sort-work-view to get the tabName. 

If you update ENavTab to include sort-work, then you can use that instead of tabName from sort-work-view.

scytacki commented on src/components/document/sort-work-document-area.tsx:
```diff
@@ -0,0 +1,91 @@
+  const problemStore = useProblemStore();
+  const appConfigStore = useAppConfig();
+
+  const navTabSpec = appConfigStore.navTabs.getNavTabSpec(tabName);
+  const tabState = navTabSpec && persistentUI.tabs.get(navTabSpec?.tab);
+  const openDocumentKey = tabState?.openDocuments.get(tabName) || "";
```
Given that SortWorkView is basically already figuring out the openDocumentKey, you might as well pass it in as a property to SortWorkDocumentArea. This way SortWorkDocumentArea doesn't have to use `appConfigStore.navTabs`, or `persistentUI`.

dennisrcao commented:
@scytacki Okay I just talked with Tejal and also wrote in the new test spec file` teacher_sort_Work_view_spec.js` some to do items for QA.
For now I added a basic skeleton test structure and believe I addressed all your points.

scytacki commented:
I added the run regression label to see how this does with the cypress tests...

scytacki reviewed (changes requested):
This looks good. The only thing you missed was merging in `186506848-sort-by-name-and-group` dealing with any merge conflicts, and setting that as the base branch. 

That branch's PR has been approved by QA though: https://github.com/concord-consortium/collaborative-learning/pull/2113

So you could just merge that PR, and merge master into this branch.

scytacki commented on src/components/document/sort-work-view.tsx:
```diff
@@ -96,10 +116,19 @@ export const SortWorkView: React.FC = observer(function SortWorkView() {
   };
 
+  const appConfigStore = useAppConfig();
+  const navTabSpec = appConfigStore.navTabs.getNavTabSpec(ENavTab.kSortWork);
+  const tabState = navTabSpec && persistentUI.tabs.get(navTabSpec?.tab);
```
```suggestion
  const tabState = navTabSpec && persistentUI.tabs.get(ENavTab.kSortWork);
```

scytacki commented on src/components/document/sort-work-view.tsx:
```diff
@@ -67,13 +69,24 @@ export const SortWorkView: React.FC = observer(function SortWorkView() {
     };
     const aParsed = parseName(a);
     const bParsed = parseName(b);
+    if (aParsed.isNumericLastName && bParsed.isNumericLastName) {
+      return aParsed.lastNameNum - bParsed.lastNameNum;
+    }
+    if (aParsed.isNumericLastName) return -1;
+    if (bParsed.isNumericLastName) return 1;
```
We shouldn't do anything special for numeric last names. Only demo users have numeric last names, if a real user has a numeric last name we shouldn't treat it specially.

scytacki reviewed (changes requested):
I found a couple of things to fix.

scytacki reviewed (approved):
looks good. 

I went ahead and just fixed one remaining issue.

tejal-shah reviewed (approved):
Looks good 👍
